packages/selection/src/useSelection.js

```diff
@@ -0,0 +1,270 @@
+/**
+ * Copyright Zendesk, Inc.
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import { useEffect, useContext, useReducer } from 'react';
+
+import { LocaleContext } from './components/LocaleProvider';
+import { composeEventHandlers } from './utils/composeEventHandlers';
+import { getControlledValue } from './utils/getControlledValue';
+import { KEY_CODES } from './utils/KEY_CODES';
+import { DIRECTIONS } from './utils/DIRECTIONS';
+import { ACTIONS } from './utils/ACTIONS';
+
+/* eslint-disable indent */
+function stateReducer(state, action, { focusedItem, selectedItem, onFocus, onSelect }) {
+  const controlledFocusedItem = getControlledValue(focusedItem, state.focusedItem);
+  const controlledSelectedItem = getControlledValue(selectedItem, state.selectedItem);
+  const currentItemIndex =
+    controlledFocusedItem === undefined
+      ? action.items.indexOf(controlledSelectedItem)
+      : action.items.indexOf(controlledFocusedItem);
+
+  switch (action.type) {
+    case ACTIONS.FOCUS:
+      if (onFocus) {
+        onFocus(action.payload);
+
+        return state;
+      }
+
+      return { ...state, focusedItem: action.payload };
+    case ACTIONS.INCREMENT: {
+      const newFocusedItem = action.items[(currentItemIndex + 1) % action.items.length];
+
+      if (onFocus) {
+        onFocus(newFocusedItem);
+
+        return state;
+      }
+
+      return { ...state, focusedItem: newFocusedItem };
+    }
+    case ACTIONS.DECREMENT: {
+      const newFocusedItem =
+        action.items[(currentItemIndex + action.items.length - 1) % action.items.length];
+
+      if (onFocus) {
+        onFocus(newFocusedItem);
+
+        return state;
+      }
+
+      return { ...state, focusedItem: newFocusedItem };
+    }
+    case ACTIONS.HOME:
+      if (onFocus) {
+        onFocus(action.items[0]);
+
+        return state;
+      }
+
+      return { ...state, focusedItem: action.items[0] };
+    case ACTIONS.END:
+      if (onFocus) {
+        onFocus(action.items[action.items.length - 1]);
+
+        return state;
+      }
+
+      return { ...state, focusedItem: action.items[action.items.length - 1] };
+    case ACTIONS.MOUSE_SELECT:
+      // eslint-disable-next-line no-case-declarations
```

Comment:
I finally found the meaning for this one... we need `{ ... }` wrappers to stop these from leaking into the higher scope.